components/log-viewer-webui/server/src/plugins/MongoReplicaServerPlugin.ts

```diff
@@ -0,0 +1,272 @@
+ *  offline; maybe add a ref count then), performance, and collision chances.
+ *
+ * @param {string} query
+ * @param options
+ * @return {string}
+ */
+const getQueryHash = (query: object, options: object): string => JSON.stringify({query, options});
+
+
+class MongoReplicaServerCollection {
+    private count: number;
+
+    private collection: any;
+
+    private watchers: Map<string, any>;
+
+    constructor(mongoDb: any, collectionName: string) {
+        this.count = 0;
+        this.collection = mongoDb.collection(collectionName);
+        this.watchers = new Map();
+    }
+
+    /**
+     * Increment the reference count;
+     */
+    refAdd () {
+        this.count++;
+    }
+
+    /**
+     * Decrement the reference count;
+     */
+    refRemove () {
+        this.count--;
+    }
+
+    /**
+     * Check if the collection is being referenced.
+     *
+     * @return {boolean}
+     */
+    isReferenced (): boolean {
+        return 0 < this.count;
+    }
+
+    find (query: object, options: object) {
+        return this.collection.find(query, options);
+    }
+
+    getWatcher (query: object, options: object) {
+        const queryHash = getQueryHash(query, options);
+        let watcher = this.watchers.get(queryHash);
+        if ("undefined" === typeof watcher) {
+            watcher = this.collection.watch({ $match: query });
+            this.watchers.set(queryHash, watcher);
+        }
+        return { queryHash, watcher };
+    }
+
+    removeWatcher (queryHash: string) {
+        if (this.watchers.has(queryHash)) {
+            this.watchers.get(queryHash).close();
+            this.watchers.delete(queryHash);
+        }
+    }
+}
+
+class MongoReplicaServer {
+    private fastify: FastifyInstance;
+
+    private collections: Map<string, MongoReplicaServerCollection>;
+
+    private mongoDb: any;
+
+
+    constructor ({fastify, mongoDb}: {fastify: FastifyInstance; mongoDb: any}) {
+        console.log('hello world')
+        this.fastify = fastify;
+        this.collections = new Map();
+        this.mongoDb = mongoDb;
+        this.initializeSocketServer(fastify.server);
+    }
+
+    static async create ({
+        fastify,
+        dbName,
+        mongoUri,
+    }: {
+        fastify: FastifyInstance;
+        dbName: string;
+        mongoUri: string;
+    }): Promise<MongoReplicaServer> {
+        const mongoDb = await MongoReplicaServer.initializeMongoClient({ mongoUri, dbName });
+
+        return new MongoReplicaServer({fastify, mongoDb});
+    }
+
+    static async initializeMongoClient (
+        {mongoUri, dbName}: {mongoUri: string; dbName: string}
+    ): Promise<any> {
+        const mongoClient = new MongoClient(mongoUri);
+        try {
+            await mongoClient.connect();
+
+            return mongoClient.db(dbName);
+        } catch (e) {
+            throw new Error("MongoDB connection error", {cause: e});
+        }
+    }
+
+    /* eslint-disable max-lines-per-function */
```

Comment:
I understand this was ported from the prototype. Let's remove this and try to deduplicate the `socket` event registrations by iterating through a list of objects like
```
{eventName: "disconnect", listener: this.#socketDisconnectListener, args: null, callback: null},
...
```